File: platforms/android/lib/src/main/java/com/shopify/checkoutkit/Configuration.kt
package com.shopify.checkoutkit

import android.content.Context

/**
 * Configuration for Shopify Checkout Kit.
 *
 * Allows specifying the colors, sheet presentation, and runtime behavior that should be used for checkout.
 *
 * @property allowedMessageOrigins Extra origins allowed to post incoming checkout-protocol messages.
 * Native checkout is open by default: leaving this empty trusts every origin. Once populated, the
 * effective allowlist is these origins plus the cart URL origin and `shop.app` (including its
 * subdomains). Entries may be exact origins (`https://example.com`), scheme-qualified wildcard
 * subdomains (`https://&#42;.example.com`), or `"*"` to explicitly trust every origin.
 * Messages dropped by origin validation are logged as warnings.
 */
@ConsistentCopyVisibility
public data class Configuration internal constructor(
    var appearance: CheckoutAppearance = CheckoutAppearance.Storefront(),
    var sheet: CheckoutSheetOptions = CheckoutSheetOptions(),
    var platform: Platform? = null,
    var logLevel: LogLevel = LogLevel.WARN,
    var preloading: Preloading = Preloading(),
    var title: String? = null,
    var allowedMessageOrigins: Set<String> = emptySet(),
)

/**
 * Resolves the checkout sheet header title, preferring a runtime-configured title over the localized default.
 */
internal fun Configuration.resolveCheckoutTitle(context: Context): String =
    title ?: context.getString(R.string.checkout_web_view_title)

public data class Preloading(
    public val enabled: Boolean = true,
)

public enum class LogLevel {
    DEBUG, WARN, ERROR, NONE
}

public sealed class Platform(
    public val identifier: String,
    public val version: String? = null,
) {
    public class ReactNative @JvmOverloads constructor(
        version: String? = null
    ) : Platform("ReactNative", version)
}

File: platforms/swift/Sources/ShopifyCheckoutKit/Configuration.swift
import UIKit

public struct Platform: Equatable, Sendable {
    public let identifier: String
    public let version: String?

    public static let reactNative = Platform(identifier: "ReactNative", version: nil)

    public static func reactNative(version: String) -> Platform {
        Platform(identifier: "ReactNative", version: version)
    }
}

public struct Configuration: Sendable {
    /// Determines the appearance used when checkout is presented.
    ///
    /// By default, checkout uses the storefront's web checkout branding.
    /// Use `.app(.automatic)`, `.app(.light)`, or `.app(.dark)` to use the
    /// Checkout Kit style instead.
    public var appearance = Appearance.storefront

    public var confetti = Configuration.Confetti()

    public var preloading = Configuration.Preloading()

    public var tintColor: UIColor = .init(red: 0.09, green: 0.45, blue: 0.69, alpha: 1.00)

    @available(*, renamed: "tintColor", message: "spinnerColor has been superseded by tintColor")
    public var spinnerColor: UIColor = .init(red: 0.09, green: 0.45, blue: 0.69, alpha: 1.00)

    public var backgroundColor: UIColor = .systemBackground

    public var logger: Logger = NoOpLogger()

    public var title: String = NSLocalizedString("shopify_checkout_kit_title", value: "Checkout", comment: "The title of the checkout sheet.")

    /// The tint color for the close button. If nil, uses the system default.
    public var closeButtonTintColor: UIColor?

    /// Custom enum for identifying traffic from alternative platforms
    public var platform: Platform?

    /// Levels: debug, warn, error, none (ordered threshold, most to least verbose)
    /// Default: .warn - which emits warnings and errors
    public var logLevel: LogLevel = .warn

    /// Origins that are trusted to send incoming checkout messages, in addition
    /// to the loaded checkout origin and shop.app.
    ///
    /// The native surface is open by default: when this is empty, messages from
    /// any origin are accepted. Provide one or more origins to restrict which
    /// origins are trusted; the loaded checkout origin and shop.app are always
    /// appended. Use `"*"` to explicitly disable origin validation.
    ///
    /// Entries are origin patterns:
    /// - `"https://example.com"` — an exact origin.
    /// - `"https://*.example.com"` — any subdomain of `example.com`.
    /// - `"*"` — allow all origins (escape hatch).
    ///
    /// An optional trailing slash is accepted. Credentials, paths, queries,
    /// and fragments are not valid in configured origin patterns.
    public var allowedMessageOrigins: [String] = []

    /// Invoked when an incoming checkout message is rejected during origin
    /// validation. Defaults to logging a debug message; rejected messages are
    /// never silently dropped.
    public var onMessageRejected: (@Sendable (MessageRejection) -> Void)?
}

extension Configuration {
    public enum ColorScheme: String, CaseIterable, Sendable {
        /// Uses a light, idiomatic color scheme.
        case light
        /// Uses a dark, idiomatic color scheme.
        case dark
        /// Infers either `.light` or `.dark` based on the current `UIUserInterfaceStyle`.
        case automatic
    }
}

extension Configuration {
    public enum Appearance: Equatable, Sendable {
        /// Uses the Checkout Kit style with the provided color scheme.
        case app(ColorScheme = .automatic)
        /// Uses the storefront's web checkout branding with a light color scheme.
        case storefront
    }
}

extension Configuration {
    public struct Confetti: Sendable {
        public var enabled: Bool = false

        public var particles = [UIImage]()
    }
}

extension Configuration {
    public struct Preloading: Sendable {
        public var enabled: Bool = true
    }
}
